kinesisfirehose-logs-extension-demo/README.md

```diff
@@ -0,0 +1,226 @@
+* automatically instrumenting your code without needing code changes
+* fetching configuration settings or secrets before the function invocation
+* detecting and alerting on function activity through hardened security agents, which can run as separate processes from the function
+
+read more about it [here](https://aws.amazon.com/blogs/compute/introducing-aws-lambda-extensions-in-preview/)
+
+> Note: The code sample provided part of this pattern uses **external** extension to listen to log events from the Lambda function
+
+## Need to centralize log collection
+
+Having a centralized log collecting mechanism using Kinesis firehose provides the following benefits:
+
+* Helps to collect logs from different sources in one place. Even though the sample provided sends logs from Lambda, log routers like `Fluentbit` and `Firelens` can send logs directly to Kinesis Data firehose from container orchestrators like `EKS` and `ECS`.
+* Define and standardize the transformations before the log gets delivered to downstream systems like S3, elastic search, redshift, etc
+* Provides a secure storage area for log data, before it gets written out to the disk. In the event of machine/application failure, we still have access to the logs emitted from the source machine/application
+
+## Architecture
+
+### AWS Services
+
+* Amazon Lambda
+* Amazon Lambda extension
+* Amazon Kinesis Data Firehose
+* Amazon S3
+
+### High level architecture
+
+Here is the high level view of all the components
+
+![architecture](images/centralized-logging.svg)
+
+Once deployed the overall flow looks like below:
+
+* On start-up, the extension subscribes to receive logs for `Platform` and `Function` events.
+* A local HTTP server is started inside the external extension which receives the logs.
+* The extension also takes care of buffering the recieved log events in a synchronized queue and writing it to AWS Kinesis Firehose via direct `PUT` records
+
+> Note: Firehose stream name gets specified as an environment variable (`AWS_KINESIS_STREAM_NAME`)
+
+* The Lambda function won't be able to send any logs events to Amazon CloudWatch service due to the following explicit `DENY` policy:
+
+```yaml
+Sid: CloudWatchLogsDeny
+Effect: Deny
+Action:
+  - logs:CreateLogGroup
+  - logs:CreateLogStream
+  - logs:PutLogEvents
+Resource: arn:aws:logs:*:*:*
+```
+
+* The Kinesis Firehose stream configured part of this sample sends log directly to `AWS S3` (gzip compressed).
+
+## Build and Deploy
+
+AWS SAM template available part of the root directory can be used for deploying the sample lambda function with this extension
+
+### Pre-requistes
+
+* AWS SAM CLI needs to get installed, follow the [link](https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/serverless-sam-cli-install.html) to learn how to install them
+
+### Build
+
+Check out the code by running the following command:
+
+```bash
+mkdir aws-lambda-extensions && cd aws-lambda-extensions
```

Comment:
Do you need to make the directory, the clone will create it anyway?

Reply:
Removed extra mkdir command